### ISSUE TITLE
Implement edition-gated keywords

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust#49610]
We want to reserve a bunch of keywords in the 2018 edition. There's a [full list of potential keywords](https://internals.rust-lang.org/t/keywords-to-reserve-for-rust2018-epoch/6720), but for now I'll probably just reserve async, await, and catch.

**EDIT:** Here's my preliminary review of potential keywords and stuff: https://hackmd.io/lu7wxNbRTuO3C0NITUum6Q?both# // @Centril

For edition hygiene reasons this will be done entirely at the lexer level. Keywords in Rust are all lexed as identifiers, however you can have "raw identifiers" which let you create idents treated explicitly as idents. We overload this, so that if the lexer comes across the identifier "async" on Rust 2015, it lexes it as if it were `r#async`. We may want to fix it up in the pretty printer as well.

**Comment by Manishearth** [comment]
An alternate fix is to have a second flag on `Ident`, which marks whether or not it's allowed to form a keyword. This has the additional benefit of being easier to work with for promoting contextual keywords to full keywords.

**Comment by Manishearth** [comment]
Actually, not doing `catch` yet, it's not contextual.

**Comment by Manishearth** [comment]
Ah, `catch` is currently implemented in an unstable form as `do catch`. We can change this whenever we like so that it doesn't work for the 2015 epoch and in 2018 uses a non-contextual catch keyword.

**Comment by Manishearth** [comment]
https://github.com/rust-lang/rust/pull/49611

**Comment by cramertj** [comment]
Why do `async` and `await` need to be non-contexutal keywords? AFAIK at least `async` could be added as a backwards-compatible contextual keyword.

**Comment by Manishearth** [comment]
Because I picked the two keywords I knew we definitely wanted 😄 (most of the others are under discussion or otherwise iffy).

I wasn't aware we were fine with them as contextual keywords -- I thought we wanted to epochify anything that we may need as a keyword in the next epoch.

Personally I'd rather add things as full keywords. But thinking about it I guess it would be nice if Rust 2015 got async/await as well.

However, AFAICT await will have to be a real keyword? Or it will be prone to the same issue we have with `catch` (which causes `do catch`)? And if `await` is going to be a 2018-only keyword it seems like the entire feature should just be 2018-only, in which case, why not make async a full keyword.

idk 😄

**Comment by cramertj** [comment]
> why not make async a full keyword

Because I have preexisting modules named `async` :stuck_out_tongue: lol

**Comment by Manishearth** [comment]
`r#async` all the way 😛

**Comment by Manishearth** [comment]
@nikomatsakis thoughts on what we should be doing here?

As mentioned in Gitter, we can do this for `dyn` as well -- no `dyn` at all in 2015, and `dyn` becomes a full keyword in 2018. I kinda prefer that.

cc @Centril

**Comment by Centril** [comment]
@Manishearth 

> [..] we can do this for dyn as well -- no dyn at all in 2015, and dyn becomes a full keyword in 2018. I kinda prefer that.
>
> cc @Centril

Yes, this ^. All the way. I think we should make switching to Rust 2018 as enticing as possible, and reduce technical debt if we can. So I'm 👍  for no `dyn` in edition 2015.

On `async`: @rpjohnst's [pre-RFC](https://internals.rust-lang.org/t/pre-rfc-await-generators-directly/7202) has `async { .. }` blocks; and so this would conflict with `struct async { .. } [..] async { .. }` wherefore I think `async` needs to be a real keyword so that we can future proof things.

On `catch`, the keyword reservation does not seem settled, but whatever we go with (`try`, `catch`, `trap`, .. ) it should be a real keyword for the same problem that `async` blocks would have.

On `await` I have no idea.

**Comment by rpjohnst** [comment]
(I intentionally didn't think too hard about syntax. `async { .. }` was just shorthand for `async || { .. }` because the arguments were always empty- so unless we actually expect that syntax to stick around for some other reason it's not super relevant.)

**Comment by Manishearth** [comment]
@rpjohnst what about await? Are either of these in a situation where we want a non-contextual keyword?

**Comment by rpjohnst** [comment]
The proposal I wrote doesn't use the `await` keyword at all, but I have no idea if that will actually work out- like the `async` case it wasn't really about the syntax.

**Comment by nikomatsakis** [comment]
Tagging this as A-rust-2018-preview. The main work remaining here, afaik, is to issue a lint when people use identifiers that will become keywords in the future. I've been debating how to implement that. I can see three basic approaches so far:

- Walk the AST for definitions/references to identifiers with a suspicious name.
- Walk the HIR for definitions/references to identifiers with a suspicious name.
- Modify the parser so that when it sees a given identifier, it records the locations + spans. 

Modifying the parser would be harder to tie in with a lint, in that we lose the "scope" information we typically use to decide if the lint is enabled, but we could say that you have to put the migration lint at the root of the crate for it to work, and I'd be ok with that.

The first two seem like they might be easier though. Have to find the right time to do it. We might be able to leverage the existing folders etc to visit every keyword, though, which is very nice. If we do it on the AST, I guess we might be able to intercept macro-related things, but I we actually don't have to do that, since AST matching (e.g., `$t:ident` etc) doesn't really care about this... right?

(Due in part to our behavior in https://github.com/rust-lang/rust/issues/49520, which I was concerned about, but now I think I see why @petrochenkov preferred it.)

cc @petrochenkov @Manishearth — I know both of you have put some thought into this, would appreciate your feedback.

**Comment by Manishearth** [comment]
A tricky bit is handling keywords in macro definitions and invocations, but aside from that i think walking the tree is fine

**Comment by nikomatsakis** [comment]
@Manishearth 

In what way is that tricky, say more =) 

I guess the idea is if we use a macro to parse something as an expression and then drop it on the floor?

e.g.,

```rust
macro_rules! validate_expr {
}
```

**Comment by nikomatsakis** [comment]
(Note: in discussion today, we decided this wasn't a *preview* blocker per se, though it is an *edition* blocker.)

**Comment by petrochenkov** [comment]
@nikomatsakis 
I think we can visit all idents in AST as soon as NodeIds are assigned and it will be good enough.

This visiting would include right sides of macro definitions which are possibly never expanded in the current crate, but will start generating errors in other crates once the edition of this crate is bumped.
They would be visited as token streams because we don't know anything more about them. Left sides don't need to be checked as you've mentioned above.

There can be false positives in corner cases, e.g. left sides of nested macros defined inside right side of a macro. Regarding the `validate_expr` example, I'm not even sure it would be a false positive - it certainly wouldn't hurt to rename `try` into something else in it, in case `validate_expr` stops dropping arguments.

**Comment by scottmcm** [comment]
FWIW, `try` is made an edition keyword by PR https://github.com/rust-lang/rust/pull/52602.

**Comment by nikomatsakis** [comment]
Does anyone have any idea what work remains to be done here?

We are now linting for the use of 2018 keywords in Rust 2015, for example, at least afaik. Based on the code I see four keywords, and all of them appear to have at least rudimentary tests:

- `async` ([test](https://github.com/rust-lang/rust/blob/5de52817c43fcb767ab3c7ecf051872cad6d0df2/src/test/ui/rust-2018/async-ident.rs))
- `dyn` ([test](https://github.com/rust-lang/rust/blob/master/src/test/ui/rust-2018/dyn-keyword.rs))
- `try` ([test as ident](https://github.com/rust-lang/rust/blob/5de52817c43fcb767ab3c7ecf051872cad6d0df2/src/test/ui/rust-2018/try-ident.rs), [test as macro](https://github.com/rust-lang/rust/blob/5de52817c43fcb767ab3c7ecf051872cad6d0df2/src/test/ui/rust-2018/try-macro.rs))
- `await` ([migrate to `r#await` in 2105](https://github.com/rust-lang/rust/blob/5de52817c43fcb767ab3c7ecf051872cad6d0df2/src/test/ui/await-keyword/2015-edition-warning.rs), [error in 2018](https://github.com/rust-lang/rust/blob/5de52817c43fcb767ab3c7ecf051872cad6d0df2/src/test/ui/await-keyword/2018-edition-error.rs), [non-error in 2018 with feature gate](https://github.com/rust-lang/rust/blob/5de52817c43fcb767ab3c7ecf051872cad6d0df2/src/test/ui/await-keyword/2018-edition-no-error-with-feature-gate.rs))

So maybe we can close this?

**Comment by Centril** [comment]
Yes I think we can close this; I filed https://github.com/rust-lang/rust/issues/54774 for improved diagnostics, but that is tracked separately.
